Realign accessible color logic around new env var

andyfeller opened the pull request:
Relates github/cli#830

1. Updates existing accessibility detection logic specifically around accessible colors
2. Updates to testing regarding changes
3. Various other enhancements to related code and comments

andyfeller committed:
Realign accessible color logic around new env var

1. Introduce constant for `gh config` setting for accessible colors
2. Align existing constant for env var to match
3. Rename exported functions to distinguish use specifically for color
4. Minor enhancements to comments and tests

williammartin commented on pkg/accessibility/accessibility.go:
```diff
@@ -6,24 +6,31 @@ import (
+// Any other value enables the accessibility features.
+//
+// Note that this environment variable supercedes the configuration file's accessible setting.
+func IsAccessibleColorsEnabled() bool {
```
Since this is living in `go-gh`, I think I'd like to see this function doc-ed as experimental and shouldn't be used by anything other than `cli/cli`.

williammartin commented:
Does this really "fix https://github.com/github/cli/issues/830"? Maybe it won't close the issue because it's cross org but it seems to me we need to either merge this to `main` and release or point `cli/cli` to this commit right? Is there any other work required beyond that?

andyfeller commented:
> Does this really "fix [github/cli#830](https://github.com/github/cli/issues/830)"?

No, good call out; I've updated this to "relates" because your next comment is on point: we need to start using it ...

> Maybe it won't close the issue because it's cross org but it seems to me we need to either merge this to `main` and release or point `cli/cli` to this commit right? Is there any other work required beyond that?

... I think we have potentially 3 choices of how to move forward:

1. **Merge this into default branch and release it**

   If we did this now as-is, then any `markdown` package user who wanted to update `cli/go-gh` would be required to do a light/moderate bit of heavy lifting to switch over to `cli/glamour`.  

   They will also have to undo it later when we finally contribute our fork changes upstream and remove `cli/glamour`; not a great experience.

1. **Point `cli/cli` to this commit**

   This would allow us begin using it without affecting other `cli/go-gh` users.

   It also means that we have to be vigilant about other `cli/go-gh` changes going into the default branch and keeping `accessible-colors` up to date until we can get the upstream changes accepted.

1. **Move `accessible-colors` work under `x/markdown` package separate of `markdown`**

   Your comment about about marking or signaling that functions are experimental reminded me of [`golang.org/x`](https://pkg.go.dev/golang.org/x) of [`charmbracelet/x`](https://github.com/charmbracelet/x).

   The idea being that this could:
   
   1. Merge this into the default branch and release
   1. Avoid impacting `cli/go-gh` users of `markdown`
   1. Reduce some of our toil iterating on `cli/cli` and `cli/go-gh` while getting changes accepted upstream
   1. Communicate status of these new changes, maybe even help to get feedback from those interested who pick up on them

andyfeller commented:
## Post-discussion with Will about the above points

1. We think `cli/glamour` changes should go into upstream before merging `accessible-colors` to avoid disrupting GitHub CLI extension authors
   - https://github.com/charmbracelet/glamour/pull/395

1. We think expanding color capabilities for GitHub CLI and CLI is at odds with `accessibility` package and want to do the following
   - [x] Rename the `accessibility` package to `color`

1. We want to ensure these changes are understood to be experimental and will do the following
   - [x] Add in-line comments about the experimental status of the accessible color logic
   - [x] Move the `color` package to `x/color` to communicate experimental status

### Reasoning

When our efforts to build a more accessible GitHub CLI experience started, we didn't know the full scope of how it would play out in `cli/cli` and `cli/go-gh` changes. We knew there were concerns around colors in general, colors in markdown, and screen readers working with prompting. As we have made progress, the `accessibility` package feels the wrong way to organize this new behavior versus what it is affecting.

The question is where does this live?

- `config`: underlying `gh config` and guts of GitHub CLI configuration
- `prompter`: interactive user prompting
- `term`: terminal environment and capability handling

The idea of a new `color` package was motivating bringing [GitHub CLI color handling logic out of `cli/cli`](https://github.com/cli/cli/blob/fc19ff321a4b7f4198f788f08da20249c7950fbd/pkg/iostreams/color.go) and into `cli/go-gh` while expanding on ideas from [Primer Design System and color roles](https://primer.style/foundations/color/overview#color-roles).

andyfeller committed:
Reorganize and communicate experimental status

Following Go `x` convention seen across various projects, the former `accessibility` package has been relocated to communicate its experimental status more clearly to extension authors.

It has also been renamed to `color` to align with adopting Primer Style color role functionality.

andyfeller committed:
Fix bugs from package renaming

andyfeller commented:
## Testing these changes against upstream `charmbracelet/glamour` contribution

@williammartin : I just finished building `cli/cli` locally against a modified version of this branch with https://github.com/charmbracelet/glamour/pull/395 contributions to verify the work within `accessible-colors`.  Once this contribution is merged, I'm confident we can switch off our fork and finish releasing these changes.

**Is there anything else you'd like to see before finalizing review here?**

**Test Branch:** https://github.com/cli/go-gh/tree/andyfeller/gh_accessible_colors_upstream
**Test Repo:** https://github.com/andyfeller/markdown-lorem-ipsum

### Demonstrating non-codeblock accessible colors

In the following screenshot, the lefthand terminal displays the current `gh` markdown color behavior of 8-bit color with the righthand terminal using this branch's changes and displays 4-bit color:

![Screenshot of rendering `andyfeller/markdown-lorem-ipsum` markdown using existing 8-bit experience focused on non-codeblocks](https://github.com/user-attachments/assets/27f74fb0-abd8-40df-bd6c-96ae3d24d5ae)

The righthand terminal's markdown headers can now be modified through terminal preferences to render magenta as canary yellow:

![Screenshot of rendering `andyfeller/markdown-lorem-ipsum` markdown using new 4-bit experience focused on non-codeblocks](https://github.com/user-attachments/assets/5862612f-dca4-4233-b964-2d68c5023a0d)

### Demonstrating codeblock accessible colors

In the following screenshot, the lefthand terminal displays the current `gh` markdown color behavior of 8-bit color with the righthand terminal using this branch's changes and displays 4-bit color:

![Screenshot of rendering `andyfeller/markdown-lorem-ipsum` markdown using existing 8-bit experience focused on codeblocks](https://github.com/user-attachments/assets/7792d423-d3f1-4684-bc3f-31bcc2d70340)

The righthand terminal's codeblocks can now be modified through terminal preferences to render white as electric blue:

![Screenshot of rendering `andyfeller/markdown-lorem-ipsum` markdown using new 4-bit experience focused on codeblocks](https://github.com/user-attachments/assets/6370ef5f-d5d7-4bcf-93e5-82e2850fa7a1)

andyfeller commented:
Adding a note that the upstream contribution has been merged but not released, so we could update this branch to merge into the default branch using upstream instead of merging back into `accessible_colors` using our fork.
